### PR TITLE
Add conditional linkage to only export vjp/jvp for public functions only when explicitly requested by an attribute.

### PR DESCRIPTION
**Pull request by pschuh**
This is based off of a PR by rxwei.

### CODE REVIEW

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
Yay!

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -5169,9 +5171,16 @@ DifferentiationTask::DifferentiationTask(ADContext &context,
           original, attr, vjpName, AutoDiffAssociatedFunctionKind::VJP);
     attr->setVJPName(vjpName);
   }
+  // If differentiation is triggered by `@differentiable`, we export its symbol
+  // when the original function is public.
+  auto isAssocFnExported =
+      invoker.getKind() ==
+          DifferentiationInvoker::Kind::DifferentiableAttribute ||
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Is it the case that `DifferentiableAttribute` is never used now?

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (pschuh) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Yes, I can send a followup to delete DifferentiableAttribute from everything tomorrow.

Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: great

**Pull request merged**